File: 2023/day16/zig/solution.zig
const std = @import("std");
const ArrayList = std.ArrayList;
const Allocator = std.mem.Allocator;

const Direction = enum(u2) {
    right = 0,
    down = 1,
    left = 2,
    up = 3,
};

const State = struct {
    row: i32,
    col: i32,
    dir: Direction,
};

// Direction deltas: right, down, left, up
const DR = [4]i32{ 0, 1, 0, -1 };
const DC = [4]i32{ 1, 0, -1, 0 };

fn countEnergized(allocator: Allocator, grid: []const []const u8, start_row: i32, start_col: i32, start_dir: Direction) !usize {
    const rows: i32 = @intCast(grid.len);
    const cols: i32 = @intCast(grid[0].len);

    // Visited set: indexed by (row * cols + col) * 4 + dir
    const visited_size: usize = @intCast(rows * cols * 4);
    var visited = try allocator.alloc(bool, visited_size);
    defer allocator.free(visited);
    @memset(visited, false);

    // BFS queue
    var queue = try ArrayList(State).initCapacity(allocator, 1000);
    defer queue.deinit(allocator);

    try queue.append(allocator, State{ .row = start_row, .col = start_col, .dir = start_dir });

    while (queue.items.len > 0) {
        const state = queue.orderedRemove(0);
        const r = state.row;
        const c = state.col;
        const d = state.dir;

        // Bounds check
        if (r < 0 or r >= rows or c < 0 or c >= cols) {
            continue;
        }

        // Visited check
        const idx: usize = @intCast((r * cols + c) * 4 + @as(i32, @intFromEnum(d)));
        if (visited[idx]) {
            continue;
        }
        visited[idx] = true;

        const cell = grid[@intCast(r)][@intCast(c)];

        // Calculate next directions based on cell type
        var next_dirs: [2]?Direction = .{ null, null };
        var dir_count: usize = 0;

        switch (cell) {
            '.' => {
                next_dirs[0] = d;
                dir_count = 1;
            },
            '/' => {
                // right->up, down->left, left->down, up->right
                const mapping = [4]Direction{ .up, .left, .down, .right };
                next_dirs[0] = mapping[@intFromEnum(d)];
                dir_count = 1;
            },
            '\\' => {
                // right->down, down->right, left->up, up->left
                const mapping = [4]Direction{ .down, .right, .up, .left };
                next_dirs[0] = mapping[@intFromEnum(d)];
                dir_count = 1;
            },
            '|' => {
                if (d == .right or d == .left) {
                    // Split: up and down
                    next_dirs[0] = .down;
                    next_dirs[1] = .up;
                    dir_count = 2;
                } else {
                    next_dirs[0] = d;
                    dir_count = 1;
                }
            },
            '-' => {
                if (d == .down or d == .up) {
                    // Split: left and right
                    next_dirs[0] = .right;
                    next_dirs[1] = .left;
                    dir_count = 2;
                } else {
                    next_dirs[0] = d;
                    dir_count = 1;
                }
            },
            else => {
                next_dirs[0] = d;
                dir_count = 1;
            },
        }

        // Add next states to queue
        for (0..dir_count) |i| {
            if (next_dirs[i]) |nd| {
                const nd_idx = @intFromEnum(nd);
                try queue.append(allocator, State{
                    .row = r + DR[nd_idx],
                    .col = c + DC[nd_idx],
                    .dir = nd,
                });
            }
        }
    }

    // Count unique energized tiles
    var energized: usize = 0;
    for (0..@as(usize, @intCast(rows))) |row| {
        for (0..@as(usize, @intCast(cols))) |col| {
            const base_idx: usize = (row * @as(usize, @intCast(cols)) + col) * 4;
            if (visited[base_idx] or visited[base_idx + 1] or visited[base_idx + 2] or visited[base_idx + 3]) {
                energized += 1;
            }
        }
    }

    return energized;
}

fn part1(allocator: Allocator, grid: []const []const u8) !usize {
    return countEnergized(allocator, grid, 0, 0, .right);
}

fn part2(allocator: Allocator, grid: []const []const u8) !usize {
    const rows: i32 = @intCast(grid.len);
    const cols: i32 = @intCast(grid[0].len);
    var max_energized: usize = 0;

    // Top row, heading down
    for (0..@as(usize, @intCast(cols))) |c| {
        const energized = try countEnergized(allocator, grid, 0, @intCast(c), .down);
        max_energized = @max(max_energized, energized);
    }

    // Bottom row, heading up
    for (0..@as(usize, @intCast(cols))) |c| {
        const energized = try countEnergized(allocator, grid, rows - 1, @intCast(c), .up);
        max_energized = @max(max_energized, energized);
    }

    // Left column, heading right
    for (0..@as(usize, @intCast(rows))) |r| {
        const energized = try countEnergized(allocator, grid, @intCast(r), 0, .right);
        max_energized = @max(max_energized, energized);
    }

    // Right column, heading left
    for (0..@as(usize, @intCast(rows))) |r| {
        const energized = try countEnergized(allocator, grid, @intCast(r), cols - 1, .left);
        max_energized = @max(max_energized, energized);
    }

    return max_energized;
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    // Read input file
    const file = try std.fs.cwd().openFile("../input.txt", .{});
    defer file.close();

    const content = try file.readToEndAlloc(allocator, 1024 * 1024);
    defer allocator.free(content);

    // Parse into grid
    var lines = try ArrayList([]const u8).initCapacity(allocator, 200);
    defer lines.deinit(allocator);

    var line_iter = std.mem.splitScalar(u8, content, '\n');
    while (line_iter.next()) |line| {
        if (line.len > 0) {
            try lines.append(allocator, line);
        }
    }

    const grid = lines.items;

    const result1 = try part1(allocator, grid);
    const result2 = try part2(allocator, grid);

    std.debug.print("Part 1: {d}\n", .{result1});
    std.debug.print("Part 2: {d}\n", .{result2});
}

File: 2023/day16/rust/solution.rs
use std::collections::VecDeque;
use std::fs;

const MAX_SIZE: usize = 256;

// Directions: 0=right, 1=down, 2=left, 3=up
const DR: [i32; 4] = [0, 1, 0, -1];
const DC: [i32; 4] = [1, 0, -1, 0];

fn count_energized(grid: &[Vec<u8>], start_row: i32, start_col: i32, start_dir: usize) -> usize {
    let rows = grid.len() as i32;
    let cols = grid[0].len() as i32;

    let mut visited = [[[false; 4]; MAX_SIZE]; MAX_SIZE];
    let mut queue: VecDeque<(i32, i32, usize)> = VecDeque::new();
    queue.push_back((start_row, start_col, start_dir));

    while let Some((r, c, d)) = queue.pop_front() {
        if r < 0 || r >= rows || c < 0 || c >= cols {
            continue;
        }

        let ru = r as usize;
        let cu = c as usize;
        if visited[ru][cu][d] {
            continue;
        }
        visited[ru][cu][d] = true;

        let cell = grid[ru][cu];
        match cell {
            b'.' => {
                queue.push_back((r + DR[d], c + DC[d], d));
            }
            b'/' => {
                let nd = [3, 2, 1, 0][d];
                queue.push_back((r + DR[nd], c + DC[nd], nd));
            }
            b'\\' => {
                let nd = [1, 0, 3, 2][d];
                queue.push_back((r + DR[nd], c + DC[nd], nd));
            }
            b'|' => {
                if d == 0 || d == 2 {
                    queue.push_back((r + DR[1], c + DC[1], 1));
                    queue.push_back((r + DR[3], c + DC[3], 3));
                } else {
                    queue.push_back((r + DR[d], c + DC[d], d));
                }
            }
            b'-' => {
                if d == 1 || d == 3 {
                    queue.push_back((r + DR[0], c + DC[0], 0));
                    queue.push_back((r + DR[2], c + DC[2], 2));
                } else {
                    queue.push_back((r + DR[d], c + DC[d], d));
                }
            }
            _ => {
                queue.push_back((r + DR[d], c + DC[d], d));
            }
        }
    }

    let mut count = 0;
    for r in 0..rows as usize {
        for c in 0..cols as usize {
            if visited[r][c][0] || visited[r][c][1] || visited[r][c][2] || visited[r][c][3] {
                count += 1;
            }
        }
    }
    count
}

fn part1(grid: &[Vec<u8>]) -> usize {
    count_energized(grid, 0, 0, 0)
}

fn part2(grid: &[Vec<u8>]) -> usize {
    let rows = grid.len() as i32;
    let cols = grid[0].len() as i32;
    let mut max_energized = 0;

    // Top row, heading down
    for c in 0..cols {
        max_energized = max_energized.max(count_energized(grid, 0, c, 1));
    }

    // Bottom row, heading up
    for c in 0..cols {
        max_energized = max_energized.max(count_energized(grid, rows - 1, c, 3));
    }

    // Left column, heading right
    for r in 0..rows {
        max_energized = max_energized.max(count_energized(grid, r, 0, 0));
    }

    // Right column, heading left
    for r in 0..rows {
        max_energized = max_energized.max(count_energized(grid, r, cols - 1, 2));
    }

    max_energized
}

fn main() {
    let input = fs::read_to_string("../input.txt").expect("Failed to read input file");
    let grid: Vec<Vec<u8>> = input
        .lines()
        .filter(|line| !line.is_empty())
        .map(|line| line.as_bytes().to_vec())
        .collect();

    println!("Part 1: {}", part1(&grid));
    println!("Part 2: {}", part2(&grid));
}
